src/memory/buffer_pool.h

```diff
@@ -266,20 +287,28 @@ class BufferPool {
      */
     boost::dynamic_bitset<> m_allocationTable;
 
-    /**
-     * List of free (unallocated) pages.
-     */
-    std::list<pageId_t> m_freePages;
-
-    /**
-     * Queue of free (no page associated to them) buffers.
-     */
-    std::queue<bufferId_t> m_freeBuffers;
-
-    /**
-     * Maps pageId_t with its bufferId_t in case it is currently in the Buffer Pool. 
-     */
-    std::map<pageId_t, bufferId_t> m_bufferToPageMap;
+    struct Partition {
+        /**
+         * List of free (unallocated) pages.
+         */
+        std::list<pageId_t> m_freePages;
+    
+        /**
+         * Queue of free (no page associated to them) buffers.
+         */
+        std::queue<bufferId_t> m_freeBuffers;
+    
+        /**
+         * Maps pageId_t with its bufferId_t in case it is currently in the Buffer Pool. 
+         */
+        std::map<pageId_t, bufferId_t> m_bufferToPageMap;
+
+        /**
+         * Partition lock to isolate concurrent operations by different threads.
+         */
+        std::unique_ptr<std::shared_timed_mutex> m_lock;
```

Comment:
I think I'd would name the variable p_lock instead of m_lock, given that the unique_ptr behaves as if it was a pointer.
Also, I don't understand why you use a shared_*_mutex instead of a regular mutex, given that currently in the code you are not allowing the acquisition of the mutex in "read/shared" mode by multiple threads at the same time.

Reply:
Fixed in dc276c7